Publish public methods

n1313 opened the pull request:
This change adds (basic) support for documenting public methods of components using jsdoc tags `@public`, `@param` and `@returns`.

sapegin commented on src/rsg-components/ReactComponent/ReactComponent.js:
```diff
@@ -10,12 +11,16 @@ export default function ReactComponent({
 }) {
 	const { name, pathLine, examples } = component;
 	const { description, props } = component.props;
+	const methods = component.props.methods ? component.props.methods.filter((method) => {
```
Maybe it’d be better to filter in the loader.

sapegin replied:
We already have some methods processing in the loader and we never need private methods, so why not send only the data that’s needed by the frontend. But that’s just my opinion ;-)

sapegin commented on src/rsg-components/Methods/Methods.css:
```diff
@@ -0,0 +1,36 @@
+.table {
```
We need to extract table with most of this CSS to a separate component and use it for props and methods.

n1313 replied:
I thought of that too, but I am not sure this should be done as a part of this PR.

sapegin replied:
It introduces duplication, so that’d make sense. But it’s up to you ;-)

sapegin commented on src/rsg-components/Methods/MethodsRenderer.js:
```diff
@@ -0,0 +1,83 @@
+			<tr key={method.name}>
+				<td className={s.cell}><Code className={s.name}>{method.name}()</Code></td>
+				<td className={s.cell}>{renderParameters(method)}</td>
+				<td className={s.cell + ' ' + s.cellDesc}>
```
Better to use `classnames` instead of concatenation.

n1313 replied:
I agree, but this is how it is done in PropsRenderer which I used as a base for my PR. I decided to not deviate from the "style" of it too much. I think a classname refactoring should be done as a separate PR across the entire codebase.

sapegin replied:
Yeah, don’t worry, it’s just a minor thing ;-)

n1313 commented:
I can try to move it to the `next` branch, but I'd really like to have this change in v4 as well. Is there a way to achieve this painlessly?

sapegin commented:
Painlessly for me — send two PRs for both branches ;-) But why do you want it in 4.x? 5.x needs testing, more time we spend in 4.x, more we postpone 5.x final release. I already use it in my project without any issues but I’d like more people to try it.

n1313 commented:
I have fixed two out of three things you've commented on and fixed the broken build (very embarrassing). I would very much like to have this change in v4 because I haven't touched v5 at all and have no idea how it works, or what the migration process is, or how difficult it will be to merge this PR into `next`. I will take a look at it as soon as possible, but I can't promise anything b/c I have no clue about v5 codebase at this moment.

sapegin commented:
Thanks, merged!

I’m going to write a changelog for 5.x in the next few days (it’s going to be hard — there are 144 commits already).

n1313 commented:
Thanks!

sapegin commented:
Looks like you broke something:

![image 2017-01-09 at 6 23 35 pm](https://cloud.githubusercontent.com/assets/70067/21776043/dadb0de4-d698-11e6-8581-c88bea98bf1c.png)

n1313 commented:
That's because I'm a doofus and empty arrays are truthy.
